CalendarProject/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -0,0 +1,128 @@
+
+
+public final class FindMeetingQuery {
+    /* 
+    * Given  alist of events in one day and a request for a meeting, 
+    * returns all the timeRanges when the meetik can take place so that each attendee can attend without having anothe roverlapping event.
+    */
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        events = new ArrayList(events);
+        events.removeIf(event -> !existCommonAttendees(event.getAttendees(), request.getAttendees()));
+        List<TimeRange> unavailableTimes = getListOfTimeRanges(events);
+        List<TimeRange> reducedUnavailableTimes = getReducedListOfTimeRanges(unavailableTimes);
+        List<TimeRange> availableTimeRanges = getComplementerTimeRanges(reducedUnavailableTimes);
+        removeTooShortTimeRanges(availableTimeRanges, request.getDuration());
+        return availableTimeRanges;
+    }
+
+    /* 
+    * Removes from the list the time ranges which are shorter than minDuration.
+    */ 
+    private void removeTooShortTimeRanges(List<TimeRange> timeRanges, long minDuration) {
+        timeRanges.removeIf(timeRange -> timeRange.duration() < minDuration);
+    }
+
+    /* 
+    * Given the timeRanges in one day, this function returns the list of complementerTimeRanges that are in the same day, but are not present in timeRanges.
+    * The union of timeRanges and complementerTimeRanges is an entire day.
+    */
+    private List<TimeRange> getComplementerTimeRanges(List<TimeRange> timeRanges) {
+        List<TimeRange> complementerTimeRanges = new LinkedList<TimeRange>();
+        int lastRangeEndPoint = 0;
+        for (TimeRange timeRange : timeRanges) {
+             if (lastRangeEndPoint < timeRange.start()) {
+                 complementerTimeRanges.add(TimeRange.fromStartEnd(lastRangeEndPoint, timeRange.start(), false));
+             }
+             lastRangeEndPoint = timeRange.end();
+        }
+        if (lastRangeEndPoint < TimeRange.END_OF_DAY ) {
+            complementerTimeRanges.add(TimeRange.fromStartEnd(lastRangeEndPoint, TimeRange.END_OF_DAY, true));
+        }
+        return complementerTimeRanges;
+    }
+
+    /*
+    * Removes the duplications from the overlapping timeRanges, and merges them into one single TimeRange. 
+    * Returns a list containing a minimal number of timeRanges, which do not overlpa, but they do contain each original timeRange.
+    */
+    private List<TimeRange> getReducedListOfTimeRanges(List<TimeRange> timeRanges) {
+        List<TimeRange> reducedTimeRanges = new LinkedList<>();
+        if (timeRanges.size()==0) {
+            return reducedTimeRanges;
+        }
+        TimeRange expandableTimeRange = timeRanges.get(0);
+        TimeRange currentTimeRange;
+        for (int i = 1; i<timeRanges.size(); i++) {
+            currentTimeRange = timeRanges.get(i);
+            if (expandableTimeRange.overlaps(currentTimeRange)) { 
+                //unite this timeRange with the previous one(s)
```

Comment:
comment shoud start with space, capitalized first word and end with period `.`

Reply:
changed